base/expr.jl

```diff
@@ -371,6 +371,167 @@ macro constprop(setting, ex)
+
+# `:terminates_globally`
+
+The `:terminates_globally` settings asserts that this method will eventually terminate
+(either normally or abnormally), i.e. does not loop indefinitely.
+
+Note: The compiler will consider this a strong indication that the method will
+      terminate relatively *quickly* and may (if otherwise legal), call this
+      method at compile time. I.e. it is a bad idea to annotate this setting
+      on a method that *technically*, but not *practically*, terminates.
+
+Note: The `terminates_globally` assertion covers any other methods called by
+      the annotated method.
+
+# `:terminates_locally`
+
+The `:terminates_locally` setting is like `:terminates_globally`, except that it only
+applies to syntactic control flow *within* the annotated method. It is thus
+a much weaker (and thus safer) assertion that allows for the possibility of
+non-termination if the method calls some other method that does not terminate.
+
+Note: `terminates_globally` implies `terminates_locally`.
+
+# `:total`
+
+The `setting` combines the following other assertions:
+    - `:consistent`
+    - `:effect_free`
+    - `:nothrow`
+    - `:terminates_globally`
+and is a convenient shortcut.
+
+Note: `@assume_effects :total` is similar to `@Base.pure` with the primary
+      distinction that the :consistent-cy requirement applies world-age wise rather
+      than globally as described above. However, in particular, a method annotated
+      `@Base.pure` is always total.
+"""
+macro assume_effects(args...)
+    (consistent, effect_free, nothrow, terminates_globally, terminates_locally) =
+        (false, false, false, false, false, false)
+    for setting in args[1:end-1]
+        if isa(setting, QuoteNode)
+            setting = setting.value
+        end
+        if setting === :consistent
+            consistent = true
+        elseif setting === :effect_free
+            effect_free = true
+        elseif setting === :nothrow
+            nothrow = true
+        elseif setting === :terminates_globally
+            terminates_globally = true
+        elseif setting === :terminates_locally
+            terminates_locally = true
+        elseif setting === :total
+            consistent = effect_free = nothrow = terminates_globally = true
+        else
+            throw(ArgumentError("@assume_effects $setting not supported"))
+        end
+    end
+    ex = args[end]
+    isa(ex, Expr) || throw(ArgumentError("Bad expression `$ex` in @constprop [settings] ex"))
+    if ex.head === :macrocall && ex.args[1] == Symbol("@ccall")
+        ex.args[1] = GlobalRef(Base, Symbol("@ccall_effects"))
+        insert!(ex.args, 3, Core.Compiler.encode_effects_override(Core.Compiler.EffectsOverride(
+            consistent, effect_free, nothrow, terminates_globally, terminates_locally
+        )))
+        return esc(ex)
+    end
+    return pushmeta!(ex, :purity, consistent, effect_free, nothrow, terminates_globally, terminates_locally)
```

Comment:
I think you need `esc` here?

```suggestion
    return esc(pushmeta!(ex, :purity, consistent, effect_free, nothrow, terminates_globally, terminates_locally))
```

ATM, I get `Base.f` instead of `f`:

```julia
julia> @macroexpand Base.@assume_effects :consistent f(x) = x
:(Base.f(var"#1#x") = begin
          $(Expr(:meta, :($(Expr(:purity, true, false, false, false, false)))))
          #= REPL[1]:1 =#
          var"#1#x"
      end)

julia> Base.GIT_VERSION_INFO.commit
"2d0b7850969f67e5c3812652b4df8760565c9acf"
```

Reply:
Yes, you are correct. I found that locally earlier today, but didn't push it yet.

Reply:
@tkf I've pushed the missing commit.